### PR TITLE
Refactor of the MergeCells plugin + fixes

### PR DESCRIPTION
**Pull request by jansiegel**
Main GH issue: #4214 

#### Issues that should be fixed with this:
- #3949
- #4035
- #4138
- #1722
- #4040
- #2898
- #3903
- #3342
- #4022
- #2590
- #1635
- #2958

### CODE REVIEW

File: src/plugins/mergeCells/mergeCells.js [hunk 1]
```diff
@@ -1,452 +1,622 @@
...
-        }
+      newDelta.row = priv.lastDesiredCoords.row ? priv.lastDesiredCoords.row - currentPosition.row : newDelta.row;
+      newDelta.col = priv.lastDesiredCoords.col ? priv.lastDesiredCoords.col - currentPosition.col : newDelta.col;
 
-        if (currentSelectedRange.from.col < 0) {
-          currentSelectedRange.from.col = 0;
-        } else if (currentSelectedRange.from.col > 0 && currentSelectedRange.from.col >= totalCols) {
-          currentSelectedRange.from.col = totalCols - 1;
-        }
+      if (delta.row > 0) { // moving down
```
Comment (wszymanski) [inline src/plugins/mergeCells/mergeCells.js]: Please clear the situation with enters.

File: src/plugins/mergeCells/mergeCells.js [hunk 2]
```diff
@@ -1,452 +1,622 @@
...
+   * @param {HTMLElement} TD The cell to be modified.
+   * @param {Number} row Row index.
+   * @param {Number} col Column index.
+   * @param {String|Number} prop The prop for the current coordinates.
+   * @param {String|Number} value The cell value.
+   * @param {Object} cellProperties The current cell properties.
+   */
+  onAfterRenderer(TD, row, col, prop, value, cellProperties) {
+    let collectionInfo = this.collectionContainer.get(row, col);
+    this.dom.applySpanProperties(TD, collectionInfo, row, col);
```
Comment (wszymanski) [inline src/plugins/mergeCells/mergeCells.js]: Enter, please.

File: src/plugins/mergeCells/calculations/autofill.js [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+
+    } else {
+      direction = 'left';
+    }
+
+    return direction;
+  }
+
+  /**
+   * Snap the drag area to the farthest merged cell, so it won't clip any of the cell merged cells.
```
Comment (wszymanski) [inline src/plugins/mergeCells/calculations/autofill.js]: Please change from `so it won't clip any of the cell merged cells` to `so it won't clip other merged cells` or something like that.

File: src/plugins/mergeCells/mergeCells.js [hunk 4]
```diff
@@ -1,178 +1,488 @@
...
+  }
+
+  /**
+   * Merge cells in the provided cell range.
+   *
+   * @private
+   * @param {CellRange} cellRange Cell range to merge.
+   * @param {Boolean} [auto=false] `true` if is called automatically, e.g. at initialization.
+   * @param {Boolean} [preventPopulation=false] `true`, if the method should not run `populateFromArray` at the end, but rather return its arguments.
+   * @returns {Array|Boolean} Returns an array of [row, column, dataUnderCollection] if preventPopulation is set to true. Otherwise, returns `undefined`.
```
Comment (wszymanski) [inline src/plugins/mergeCells/mergeCells.js]: It doesn't correspond to code.

File: src/plugins/mergeCells/cellCollection/collectionContainer.js [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    });
+  }
+
+  /**
+   * Check if the provided collection overlaps with the others in the container.
+   *
+   * @param {Collection} collection The collection to check against all others in the container.
+   * @return {Boolean} `true` if the provided collection overlaps with the others, `false` otherwise.
+   */
+  checkIfOverlaps(collection) {
```
[the remark this answers is not in the document]
Comment (jansiegel) [reply src/plugins/mergeCells/cellCollection/collectionContainer.js]: Changed to `isOverlapping`

**Pull request merged**